types/wlr_seat.c

```diff
@@ -46,19 +44,69 @@ static void wl_seat_get_pointer(struct wl_client *client,
 		// the same seat
 		wl_resource_destroy(handle->pointer);
 	}
-	handle->pointer = wl_resource_create(client, &wl_pointer_interface, 5, id);
+	handle->pointer = wl_resource_create(client, &wl_pointer_interface,
+		wl_resource_get_version(_handle), id);
 	wl_resource_set_implementation(handle->pointer, &wl_pointer_impl,
-			handle, NULL);
+			handle, &wl_pointer_destroy);
+}
+
+static const struct wl_keyboard_interface wl_keyboard_impl = {
+	.release = resource_destroy
+};
+
+static void wl_keyboard_destroy(struct wl_resource *resource) {
+	struct wlr_seat_handle *handle = wl_resource_get_user_data(resource);
+	if (handle->keyboard) {
+		handle->keyboard = NULL;
+	}
 }
 
 static void wl_seat_get_keyboard(struct wl_client *client,
 		struct wl_resource *_handle, uint32_t id) {
-	wlr_log(L_DEBUG, "TODO: wl_seat_get_keyboard");
+	struct wlr_seat_handle *handle = wl_resource_get_user_data(_handle);
+	if (!(handle->wlr_seat->capabilities & WL_SEAT_CAPABILITY_KEYBOARD)) {
+		return;
+	}
+	if (handle->keyboard) {
+		// TODO: this is probably a protocol violation but it simplifies our
```

Comment:
I added this note to the other functions. It is indeed problematic, weston keeps a list of keyboard/pointer/touch resources. But this can wait for later i guess since it complicates things that have no real use case (as stated).